lib/analyse.ml

```diff
@@ -27,51 +33,102 @@ module Analysis = struct
+  
+  let check_opam_version =
+    let version_2 = OpamVersion.of_string "2" in
+    fun name opam ->
+      let opam_version = OpamFile.OPAM.opam_version opam in
+      if OpamVersion.compare opam_version version_2 < 0 then
+        Fmt.failwith "Package %S uses unsupported opam version %s (need >= 2)" name (OpamVersion.to_string opam_version)
+
+  let of_dir ~job ~master dir =
     (* TODO: Check if the PR added an opam file in packages/<pkg> instead of packages/<pkg>/<pkg>.<ver> (common mistake) *)
     (* TODO: Split modified vs. added (using git diff --name-status) *)
-    let (_, master) = Current_git.Commit.hash head in
-    let master = Option.get master in
-    let fmt = Printf.sprintf in
-    let cmd = "", [| "sh"; "-c"; fmt {|git diff %s | sed -E -n -e 's,^\+\+\+ b/packages/[^/]*/([^/]*)/.*,\1,p'|} master |] in
-    Current.Process.check_output ~cwd:dir ~cancellable:true ~job cmd >>!= fun output ->
-    let opam_files =
-      String.split_on_char '\n' output
-      |> List.filter (fun s -> not (String.equal s ""))
-      |> List.sort String.compare
-    in
-    let r = { opam_files } in
-    Current.Job.log job "@[<v2>Results:@,%a@]" Yojson.Safe.(pretty_print ~std:true) (to_yojson r);
-    Lwt.return (Ok r)
+    let master = Current_git.Commit.hash master in
+    let cmd = "", [| "git"; "merge"; master |] in
+    Current.Process.exec ~cwd:dir ~cancellable:true ~job cmd >>= function
+    | Error (`Msg msg) ->
+      Current.Job.log job "Merge failed: %s" msg;
+      Lwt_result.fail (`Msg "Cannot merge to master - please rebase!")
+    | Ok () ->
+      let cmd = "", [| "git"; "diff"; "--name-only"; master; "packages/" |] in
+      Current.Process.check_output ~cwd:dir ~cancellable:true ~job cmd >>!= fun output ->
+      let opam_files =
+        String.split_on_char '\n' output
+        |> List.filter_map (fun path ->
+            match String.split_on_char '/' path with
+            | [] | [""] | ["packages"] | ["packages"; _] -> None
+            | "packages" :: name :: package :: _ ->
+              let nme = OpamPackage.Name.of_string name in
+              let pkg = OpamPackage.of_string package in
+              if OpamPackage.Name.compare nme (OpamPackage.name pkg) <> 0 then
+                Fmt.failwith "Mismatch between package dir name %S and parent directory name %S" package name;
+              Some pkg
+            | _ ->
+              Fmt.failwith "Unexpected path %S in output (expecting 'packages/name/pkg/...')" path
+          )
+        |> List.sort_uniq OpamPackage.compare
+        |> List.map (fun pkg ->
+            let path =
+              Printf.sprintf "packages/%s/%s/opam"
+                (OpamPackage.name_to_string pkg)
+                (OpamPackage.to_string pkg)
+            in
+            (* Check it exists, parses, and is the right version. *)
+            let content = read_file ~max_len:102400 (Fpath.to_string dir / path) in
+            let opam = OpamFile.OPAM.read_from_string content in
```

Comment:
I think there should be a filter in case the file does not exist then it should be removed from the list (using `git diff --diff-filter d` is not enough because we want to check packages if only something in its `files` directory changed as well).

Also if the directory exists but the opam file does not this should be an error

Reply:
Done.